basic_autonomy/src/helper_functions.cpp

```diff
@@ -62,12 +62,12 @@ namespace waypoint_generation
         return best_index;
     }
 
-    int get_nearest_index_by_downtrack(const std::vector<lanelet::BasicPoint2d>& points, const carma_wm::WorldModelConstPtr& wm, double ending_downtrack)
+    int get_nearest_index_by_downtrack(const std::vector<lanelet::BasicPoint2d>& points, const carma_wm::WorldModelConstPtr& wm, double target_downtrack)
     {
         int best_index = points.size() - 1;
-        for(int i=0;i < points.size();i++){
+        for(int i = 0;i < points.size(); i++){
             double downtrack = wm->routeTrackPos(points[i]).downtrack;
-            if(downtrack > ending_downtrack){
+            if(downtrack > target_downtrack + epsilon_){
```

Comment:
Can you provide a description of what error this change is fixing? Seems you are accounting for floating point arithmetic errors, but since this doesn't have an == check I'm a bit unclear on why not having the epsilon is an issue

Reply:
This was resulting in an exception when the downtrack was really small (to the order of 10e-13) and still going through this loop because the target downtrack was exactly 0.0

Reply:
I guess my preference in that case would be for us to have the system return 0 when no match is found rather than throwing an exception and needing this epsilon. A point located before the input would give negative downtrack and should return 0 index, which makes sense as that would be the nearest downtrack point. Seems like that could be easily implemented by changing the default value of best_index to 0. The comments should be updated to reflect that and the fact this is not a nearest function but actually a nearest less than method.

Reply:
The way the function is written, shouldn't the best index be initialized to point.size() -1 like it currently is?
We are checking against downtrack>target_downtrack and the only time best_index will not be changed is if we go through the entire list without meeting that condition. So the nearest point would be the last one.

This would be in addition to replacing the exception for negative values with a return 0.

Reply:
Yeah I think you are right.